keps/104-StartupPolicy/README.md

```diff
@@ -0,0 +1,420 @@
+            - name: driver
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+  - name: worker
+    replicas: 2
+    template:
+      spec:
+        backoffLimit: 0 
+        completions: 2
+        parallelism: 2
+        template:
+          spec:
+            containers:
+            - name: worker
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+In the JobSet API, ReplicatedJobs are the individual Job elements that are created.  These are Replicated so you can have the same Job replicated.  This adds some complexity in the concept of Ready as we must consider the overall ReplicatedJob Ready rather than using the individual Jobs.
+
+For example, a Job with a `Parallelism`: 4 and `Ready` of 3. Means that 1 pod of a Job is not ready.  
+The Job API has no concept of an overall readiness condition based on a percentage of pods that are ready.  
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### API Proposal
+
+Keeping the non specified jobs halted we propose to use suspend for these jobs.
+
+So we will add a new field to the ReplicatedJobStatus.
+
+```golang
+// ReplicatedJobStatus defines the observed ReplicatedJobs Readiness.
+type ReplicatedJobStatus struct {
+  ...
+	Suspended int32  `json:"suspended"`
+}
+```
+
+```golang
+// Status of ReplicatedJob for StartupPolicy
+type Status string
+
+const (
+ // Ready means that the job is ready
+ Ready Status = "Ready"
+
+ +// Suspended means that the job is suspended
+ +Suspended Status = "Suspended"
+)
+
+type StartupPolicy struct {
+  // If this field is set to true, we will start all replicatedJobs in order.
+	// Started means that all the Jobs of a ReplicatedJob are ready.
+	// +optional
+	ReplicatedJobsStartInOrder *bool `json:"replicatedJobsStartInOrder,omitempty"`
+}
+```
+
+### Implementation
+
+If StartupPolicy is set and suspend field is off, we will start all jobs in a suspended state.
+
+We will not set a condition for JobSet as suspended so we can distinguish between a suspended JobSet and a JobSet with startup policy.
+To track which ReplicatedJobs are suspended, we will add a `Suspended` field to the ReplicatedJobStatus.  
+This field will track the number of replicas that are suspended.
+
+If `ReplicatedJobsStartInOrder` is set, then we will start
+We will loop over all `ReplicatedJobStatuses`.
+We will unsuspend if the ReplicatedJob is suspended and we will trigger a reconcile.
```

Comment:
```suggestion
If `ReplicatedJobsStartInOrder` is set, then we will loop over all `ReplicatedJobStatuses`.
We will start the ReplicatedJob if it is the next in order and trigger a reconcile.
```

Reply:
Wouldn't we get a reconcile anyway triggered by the previous replicatedjob updates?

Reply:
In the startup policy function, I was thinking that if a ReplicatedJob is unsuspended, we will update but we should also return and trigger the reconcile. We don't want to unsuspend the other jobs in the list.